### DISCUSSION TITLE
Optional type (?:) in autoincrement primary keys

### DISCUSSION THREAD

**Discussion by gustavonobreza** [prisma/prisma#6735] [Q&A] [answered]
## Problem
I always get frustrated when I create a new user (for example) and use the auto types (created when db is analyzed) and the id type (primary key) is not optional. Why do I need to replace the `id` value for undefined when I'm building the query.

## Suggested solution
Change the types of Primary Keys with automatic increment to optional or create a new type based on the model, but specific to create new rows in the table.

For example, I will use express. When I create a user, the `user` type imported from` @prisma/client` is being used in `req.body` (using the post HTTP method), but the client side doesn’t know what your ID for this new user and all data is passed in, but not the ID number.

#### Here I need to override for prisma auto increments ID the number

```ts
import { user } from "@prisma/client"
const Data: user = req.body 
const newRow = await addUserInDB(Data)

async function addUserInDB(userToAdd: user) {
     const addingUser = await prisma.user.create({
         data: { ...userToAdd, id: undefined }
     })
}
```
<br />

#### Here don't I need to override

```ts
import { user } from "@prisma/client"
const Data: user = req.body 
const newRow = await addUserInDB(Data)

async function addUserInDB(userToAdd: user) {
    const addingUser = await prisma.user.create({
        data: { ...UserToAdd}
    })
}
```

#### OR

```ts
import { userInput } from "@prisma/client"
const Data: userInput = req.body 
const newRow = await addUserInDB(Data)

async function addUserInDB(userToAdd: userInput) {
    const addingUser = await prisma.user.create({
        data: { ...UserToAdd}
    })
}
```

## Impact
Thus, it should be much simpler to create new records in a database table if the `PK's` with automatic increment are of an optional type. This increases performance to code and simplifies implementation.

NOTE: sorry for my bad English, I'm learning the language

**Comment by pr0ph0z** [comment, marked as answer]
May I see your schema example?

**Reply by gustavonobreza** [reply]
#### I removed some columns to make it simpler
```prisma
model User {
  id        Int      @id @default(autoincrement())
  name      String   @db.VarChar(255)
  age       Int?
  email     String?  @db.VarChar(255)
  purchases purchases[]
}
```
I'm using PostgresSQL

**Reply by pr0ph0z** [reply]
Well, I have no experience in TypeScript but I just checked the auto-generated types and see that id (which is the PK) is automatically optional by default. Or maybe the type you mean is `UserCreateInput`?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
